### ISSUE TITLE
[feature request] Runtime warning for inappropriate labels (among others)

### ISSUE THREAD

**Issue by domaala** [pytorch/pytorch#10298]
While you can get an exception when passing integers where floats are required, I believe there are no checks that actually make sure that you are using the expected **values** within your tensors. In turn, this makes tracking problems in your models a bit more challenging.

For instance, when using [`nn.CosineEmbeddingLoss`](https://pytorch.org/docs/stable/nn.html#torch.nn.CosineEmbeddingLoss), that asks for `target` to take either `1` or `-1`, nothing stops you from providing something else and experiencing weird model behaviour when the method silently interpretes it in its own way.

For example:

```
criterion = torch.nn.CosineEmbeddingLoss()

vec1 = torch.tensor([[1., 2., 3.]])
vec2 = torch.tensor([[1., 2., 3.]])

loss1 = criterion(a, b, torch.tensor([1.])) # returns 0
loss2 = criterion(a, b, torch.tensor([0.])) # returns 0
loss3 = criterion(a, b, torch.tensor([-0.5])) # returns 0
loss4 = criterion(a, b, torch.tensor([-0.99])) # returns 0
loss5 = criterion(a, b, torch.tensor([-1.])) # returns 1
```

To give an anecdote, I thought that the method takes `0` and `1` instead of `-1` and `1` and the rest is a long debugging history. :)

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1080
GPU 1: GeForce GTX 1080

Nvidia driver version: 390.30
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.5.1.5
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.6.0.21
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.14.1)
[pip3] torch (0.4.0)
[pip3] torchtext (0.2.3)
[pip3] torchvision (0.2.1)
[conda] Could not collect

**Comment by ngimel** [comment]
High priority as it is unexpected results. We check values for correctness in many other cases, we should discuss if it is one of those.

**Comment by heitorschueroff** [comment]
I think we should add a check for this parameter to be consistent.

**Comment by gchanan** [comment]
do we have a general rule for when we check?  If enough people complain?

**Comment by heitorschueroff** [comment]
@mruberry and @ngimel suggest we should come up with a consistent rule for when to check input parameters.
@mruberry should we check only on CPU or CPU/CUDA

**Comment by kshitij12345** [comment]
Implementation
https://github.com/pytorch/pytorch/blob/b9e12a0e8290af1c05107cf7e9a6117355e5ac1a/aten/src/ATen/native/Loss.cpp#L30-L44

Added check will incur perf-hit.

Similar discussion regarding adding check in https://github.com/pytorch/pytorch/issues/49998.
